compiler/src/iree/compiler/Dialect/Flow/Transforms/DumpDispatchGraph.cpp

```diff
@@ -0,0 +1,422 @@
+    AttributeMap attrs;
+    attrs["style"] = style.str();
+    // Do not label edges that start/end at a cluster boundary. Such edges are
+    // clipped at the boundary, but labels are not. This can lead to labels
+    // floating around without any edge next to them.
+    if (!n1.clusterId && !n2.clusterId)
+      attrs["label"] = quoteString(escapeString(std::move(label)));
+    // Use `ltail` and `lhead` to draw edges between clusters.
+    if (n1.clusterId)
+      attrs["ltail"] = "cluster_" + std::to_string(*n1.clusterId);
+    if (n2.clusterId)
+      attrs["lhead"] = "cluster_" + std::to_string(*n2.clusterId);
+
+    edges.push_back(strFromOs([&](raw_ostream &os) {
+      os << llvm::format("v%i -> v%i ", n1.id, n2.id);
+      emitAttrList(os, attrs);
+    }));
+  }
+
+  /// Emit a graph. The specified builder generates the body of the graph.
+  void emitGraph(function_ref<void()> builder) {
+    os << "digraph G {\n";
+    os.indent();
+    // Edges between clusters are allowed only in compound mode.
+    os << attrStmt("compound", "true") << ";\n";
+    builder();
+    os.unindent();
+    os << "}\n";
+  }
+
+  /// Emit a node statement.
+  Node emitNodeStmt(Operation *op) {
+    int nodeId = ++counter;
+    AttributeMap attrs;
+    auto label = getLabel(op);
+    auto shape = getShape(op);
+    attrs["label"] = quoteString(escapeString(std::move(label)));
+    attrs["shape"] = shape.str();
+    os << llvm::format("v%i ", nodeId);
+    emitAttrList(os, attrs);
+    os << ";\n";
+    return Node(nodeId);
+  }
+  Node emitNodeStmt(std::string label, StringRef shape = kShapeNode) {
+    int nodeId = ++counter;
+    AttributeMap attrs;
+    attrs["label"] = quoteString(escapeString(std::move(label)));
+    attrs["shape"] = shape.str();
+    os << llvm::format("v%i ", nodeId);
+    emitAttrList(os, attrs);
+    os << ";\n";
+    return Node(nodeId);
+  }
+
+  void printOperands(raw_ostream &os, ::mlir::Operation::operand_range operands,
+                     AsmState &state) {
+    for (unsigned i = 0, e = operands.size(); i != e; ++i) {
+      auto operand = operands[i];
+      auto op = operand.getDefiningOp();
+
+      if (isScalarConstantOp(op)) {
+        auto ty = operand.getType();
+        if (ty.isa<IntegerType>()) {
+          os << cast<arith::ConstantIntOp>(op).value();
+        } else if (ty.isa<FloatType>()) {
+          cast<arith::ConstantFloatOp>(op).value().print(os);
+        } else {
+          os << cast<arith::ConstantIndexOp>(op).value();
+        }
+      } else {
+        operand.printAsOperand(os, state);
+      }
+
+      if (i != e - 1) {
+        os << ", ";
+      }
+    }
+  }
+
+  /// Generate a label for an operation.
+  std::string getLabel(Operation *op) {
+    return strFromOs([&](raw_ostream &os) {
+      if (op->getNumRegions() == 0) {
+        auto funcOp = op->getParentOfType<func::FuncOp>();
+        AsmState state(funcOp);
+
+        for (auto result : op->getResults()) {
```

Comment:
I am not sure the SSA value itself is very useful (again where using record node shape helps, makes SSA value printing unnecessary). If this is just rendering after dispatch region formation, then the order of SSA values, and the SSA value name is kind of superfluous information. I'd just look at the edge to find the dependence.
Also I expect this is really only useful when used in conjunction with the IR dump. There you can search for SSA value. Since graphviz renders as a pdf the SSA value is not searchable.

Reply:
It is greatly helpful when searching a node by text. Also, along with the number on the edges, it is helpful in finding the source node when the edges are drawn in a small area and not easy to discern.